src/operator/nn/mkldnn/mkldnn_sum.cc

```diff
@@ -48,28 +49,118 @@ void Sum(const mkldnn::memory &arr1, const mkldnn::memory &arr2,
   MKLDNNStream::Get()->RegisterPrim(mkldnn::sum(sum_pd, inputs, out));
 }
 
-void MKLDNNSumForward(const nnvm::NodeAttrs& attrs, const OpContext &ctx,
-                      const std::vector<NDArray> &inputs, const OpReqType &req,
-                      const NDArray &out_data) {
-  TmpMemMgr::Get()->Init(ctx.requested[0]);
-  std::vector<mkldnn::primitive::at> in_prims;
-  std::vector<mkldnn::memory::primitive_desc> in_pds(inputs.size());
-  std::vector<float> scales(inputs.size());
+typedef MKLDNNParamOpSign<ElementWiseSumParam> MKLDNNSumSignature;
+
+class MKLDNNSumFwd {
+ public:
+  explicit MKLDNNSumFwd(const std::vector<NDArray> &inputs) {
+    _Init(inputs);
+  }
+  ~MKLDNNSumFwd() {}
+  void SetDataHandle(const std::vector<NDArray> &inputs,
+                     const NDArray &output,
+                     const OpReqType &req);
+  void Execute(const NDArray &output);
+
+ private:
+  void _Init(const std::vector<NDArray> &inputs);
+
+ private:
+  std::shared_ptr<mkldnn::sum> fwd;
+  std::vector<std::shared_ptr<mkldnn::memory>> in_data;
+  mkldnn_output_t out;
```

Comment:
you shouldn't use mkldnn_output_t. this is designed to use with CreateMKLDNNMem and CommitOutput. it's not supposed to use for holding the mkldnn memory because the second field is a raw pointer. We shouldn't use a raw pointer to hold memory.

Reply:
Thanks for the suggestion, will change to use std::shared_ptr to hold the output memory.